jsregexp.c

```diff
@@ -262,10 +269,172 @@ static int regexp_tostring(lua_State *lstate)
+
+  // check what happens in JS
+  if (r->last_index >= input_len) {
+    r->last_index = input_len;
+  }
+
+  const int capture_count = lre_get_capture_count(r->bc);
+  const int global = lre_get_flags(r->bc) & LRE_FLAG_GLOBAL;
+  const char* group_names = lre_get_groupnames(r->bc);
+
+  const bool matched = lre_exec(capture, r->bc, (uint8_t *) input,
+      global ? r->last_index : 0, input_len, 0, NULL) == 1;
+
+  if (!matched) {
+    r->last_index = 0;
+    return 0;
+  }
+
+  if (capture[0] == capture[1]) {
+    r->last_index++;
+  } else {
+    r->last_index = capture[1] - (uint8_t *) input;
+  }
+
+  lua_createtable(lstate, capture_count + 1, capture_count + 3);
+  luaL_getmetatable(lstate, JSREGEXP_MATCH_MT);
+  lua_setmetatable(lstate, -2);
+
+  lua_pushstring(lstate, input);
+  lua_setfield(lstate, -2, "input");
+
+  lua_pushnumber(lstate, 1 + capture[0] - (uint8_t *) input); // 1-based
+  lua_setfield(lstate, -2, "index");
+
+  lua_pushlstring(lstate, (char *) capture[0], capture[1] - capture[0]);
+  lua_rawseti(lstate, -2, 0);
+
+  if (group_names) {
+    lua_newtable(lstate);               // match.groups
+    lua_pushvalue(lstate, -1);
+    lua_setfield(lstate, -3, "groups"); // immediately insert into match
+    lua_insert(lstate, -2);             // leave table below the match table
+  }
+
+  for (int i = 1; i < capture_count; i++) {
+    lua_pushlstring(lstate, (char *) capture[2*i], capture[2*i+1] - capture[2*i]);
+
+    if (group_names) {
+      // if the current group is named, duplicate and insert into the correct
+      // table
+      if (*group_names) {
+        lua_pushvalue(lstate, -1);
+        lua_setfield(lstate, -4, group_names);
+        group_names += strlen(group_names);
+      }
+      group_names++;
+    }
+
+    lua_rawseti(lstate, -2, i);
+  }
+
+  return 1;
+}
+
+
+static int regexp_test(lua_State *lstate)
+{
+  uint8_t *capture[CAPTURE_COUNT_MAX * 2];
+
+  size_t input_len;
+  struct regexp *r = luaL_checkudata(lstate, 1, JSREGEXP_MT);
+  const char *input = luaL_checklstring(lstate, 2, &input_len);
+
+  const int global = lre_get_flags(r->bc) & LRE_FLAG_GLOBAL;
+
+  const bool matched = lre_exec(capture, r->bc, (uint8_t *) input,
+      global ? r->last_index : 0, input_len, 0, NULL) == 1;
+
+  if (global) {
+    if (!matched) {
+      r->last_index = 0;
+    }
+
+    if (capture[0] == capture[1]) {
+      r->last_index++;
+    } else {
+      r->last_index = capture[1] - (uint8_t *) input;
+    }
+  }
+
+  lua_pushboolean(lstate, matched);
+  return 1;
+}
+
+
+// more gettable fields to be added here
+static int regexp_index(lua_State *lstate)
+{
+  struct regexp *r = luaL_checkudata(lstate, 1, JSREGEXP_MT);
+
+  luaL_getmetatable(lstate, JSREGEXP_MT);
+  lua_pushvalue(lstate, 2);
+  lua_rawget(lstate, -2);
+
+  if (lua_isnil(lstate, -1)) {
+    const char *key = lua_tostring(lstate, 2);
+    if (streq(key, "last_index")) {
+      lua_pushnumber(lstate, r->last_index);
```

Comment:
I think `last_index` should return `lua_pushnumber(lstate, r->last_index + 1)` to keep lua 1-indexing convention